[4 Week] feature : 메인화면 부차 기능 구현

akuby21 opened the pull request:
## #️⃣연관된 이슈

#14 

## 📝작업 내용

- viewModel, useCase, repository에 해당하는 컴포넌트 구현
- 현재는 mock 데이터를 이용하여 ui 구현
- 화면의 좌, 우 커버를 눌러 노래 전환 기능
- 상단 Spinner를 통해 친구 전환 

### 스크린샷 (선택)
![image](https://github.com/user-attachments/assets/005f1adb-5a3b-4f3e-9a4a-d7198127f4ec)

## 💬리뷰 요구사항(선택)

- 아키텍처, SOLID 원칙에 어긋나는 코드가 존재하는지 알고싶습니다!
- Flow, StateFlow를 처음 사용해봐 아직 익숙치 않습니다. 올바르게 사용한지 궁금합니다
- Entity -> VO 전환 시 null 체크로 에러시 다른 값을 지정하였는데 잘한것인지 모르겠습니다. Entity 생성시 기본 값을 넣어주는 형태가 나을까요?
- viewModel 내에 화면을 그리는데 필요한 값만을 uiState로 빼서 구성하고 있습니다. 이상한 부분이 있다면 지적부탁드립니다!

akuby21 committed:
feat: add Entity for mainScreen

akuby21 committed:
feat: add VO for mainScreen

akuby21 committed:
feat: implement basic UI for mainScreen

akuby21 committed:
fix: fix type of uiState

Bitmap? -> String?

akuby21 committed:
fix: remove debug codes

akuby21 committed:
fix: add more mock datas

akuby21 committed:
fix: remove unnecessary codes

remove codes during debugging

akuby21 committed:
feat: add bindingAdapter for imageView

changing like_button when current song is like

bigstark commented on app/build.gradle.kts:
```diff
@@ -42,7 +46,8 @@ dependencies {
     implementation(libs.androidx.activity)
     implementation(libs.androidx.constraintlayout)
-    
+    implementation("androidx.fragment:fragment-ktx:1.3.0")
```
p5; version catalog 로 관리해주시면 좋지 않을까 싶네요!

akuby21 replied:
version catalog 적용했습니다!

bigstark commented on app/build.gradle.kts:
```diff
@@ -26,6 +27,9 @@ android {
         }
     }
+    buildFeatures {
+        dataBinding = true
```
p3; data binding 은 최대한 지양해주시는 편이 좋아요. 앞으로 deprecated 될 예정이기도 하구요!

akuby21 replied:
deprecated 된다는건 몰랐네요. viewBinding으로 변경 후 적용하였습니다!

bigstark commented on app/src/main/java/com/team22/soundary/feature/main/MainActivity.kt:
```diff
@@ -1,18 +1,73 @@
+
+    private fun observeUiState() {
+        lifecycleScope.launch {
+            viewModel.uiState.collect { uiState ->
+                binding.musicNameTextView.text = uiState.musicName
+                binding.singerTextView.text = uiState.singer
+                binding.messageTextView.text = uiState.message
+                binding.nextImageView.visibility = if (uiState.isLastSong) GONE else VISIBLE
+                binding.prevImageView.visibility = if (uiState.isFirstSong) GONE else VISIBLE
+            }
```
ui state 를 활용을 너무 잘하고 계신데요? 
p999) visibility 의 경우 binding.nextImageView.isGone = uiState.isLastSong 으로 처리해볼 수 있을거에요 :)

akuby21 replied:
적용했습니다! `isGone` 값은 처음 알았네요. 확실히 더 깔끔한거 같습니다

bigstark commented on app/src/main/java/com/team22/soundary/feature/main/MainActivity.kt:
```diff
@@ -1,18 +1,73 @@
     }
-}
\ No newline at end of file
+}
```
p3; 사실 이런부분들이 data binding 을 좋아하지 않는 이유기도 해요. 차라리 ui 를 처리하는 activity 에서 핸들링하는 것이 조금 더 좋아보입니다.

akuby21 replied:
저도 `bindingAdapter`를 작성하면서
'이러면 오히려 로직들이 더 흩뿌려지는게 아닌가?' 라는 생각이 스쳐지나간 것 같아요
viewBinding 변경 후 activity 내에서 핸들링하도록 변경했습니다!

bigstark commented on app/src/main/java/com/team22/soundary/feature/main/MainViewModel.kt:
```diff
@@ -0,0 +1,79 @@
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+class MainViewModel(
+    private val getShareUseCase: GetShareUseCase = GetShareUseCase()
+) : ViewModel() {
```
p3; 그나저나 hilt 를 사용한 DI 관리는 계획에 없으신걸까요?

akuby21 replied:
`hilt`이용한 DI도 당연히 도입할 예정입니다!
다만 급히 mock 데이터만 처리하다보니 아직 적용을 하지 못했네요..
이번 주 작업사항들 머지 후에 통합 브랜치에서 기본 설정 후 적용하겠습니다

bigstark commented on app/src/main/java/com/team22/soundary/feature/main/MainViewModel.kt:
```diff
@@ -0,0 +1,79 @@
+    private var groupedShares: Map<String, List<Share>> = emptyMap()
+    val friendNames: List<String>
+        get() = groupedShares.keys.toList() //for setting spinner
+    private var _currentFriendIndex = 0
+    private var _currentSongIndex = 0
```
p3; 굳이 없어도 되는 변수인 것 같아요. _uiState.value 에서 충분히 가져다 쓸 수 있을 것 같네요 :)

akuby21 replied:
노래 변경 및 친구 변경 기능이 있는 메인 UI 특성상 index 값을 써야하는데, 
현재는 UseCase로부터 `Map<String,List<Share>>`형식으로 데이터를 전달받아 편의상 뒀던 변수입니다.

피드백 주신대로 해당 변수들 대신 uiState 값을 바탕으로 인덱스를 찾는 private 함수를 작성하였습니다!

bigstark commented on app/src/main/java/com/team22/soundary/feature/main/data/ShareRepositoryImpl.kt:
```diff
@@ -0,0 +1,83 @@
+    )
+
+
+    override fun getShareList(): Flow<List<Share>> = flow {
+        emit(
+            mockDataList.map { it.toVO() }
+        )
+    }
```
p999)  아래와 같은 형태로 치환이 가능합니다 :)
```suggestion
    override fun getShareList(): Flow<List<Share>> = flowOf(mockDataList).map { it.toVO() }
```

akuby21 replied:
제안 주신대로 수정하였습니다. 훨씬 깔끔하니 좋네요!

bigstark reviewed (commented):
> 아키텍처, SOLID 원칙에 어긋나는 코드가 존재하는지 알고싶습니다!

일단 DI 에 대한 의견이 궁금해요. DI 를 사용하지 않으면 사실 안드로이드에서 SOLID 하게 코드를 짜는 것이 굉장히 어려울 수 있어요. 

> Entity -> VO 전환 시 null 체크로 에러시 다른 값을 지정하였는데 잘한것인지 모르겠습니다. Entity 생성시 기본 값을 넣어주는 형태가 나을까요?

다른값이라고 한다면 기본값일텐데, 선택이긴 합니다. 보통 해당 모델을 UI 에서 그대로 사용하는 경우가 많고 UI 는 대부분 기본 상태값들이 존재하기 때문에, 그에 맞춰서 해보면 되지 않을까 싶어요!

akuby21 committed:
refactor: change dataBinding to viewBinding

akuby21 committed:
refactor: remove unnecessary variable in ViewModel

remove friend, song Index variable.
Instead of that, viewModel will search index by getCurrentShareIndex()

akuby21 committed:
fix: fix missing uiState connection

akuby21 committed:
chore: add version catalog

akuby21 commented:
정성스러운 코드 리뷰 감사드립니다!
알고보니 이번 주는 코드 리뷰가 없었는데 제가 착각하고 슬랙에 언급드렸었네요..
괜히 멘토님 수고스럽게 만든 것 같아 죄송합니다 ㅠㅠ 다음부턴 제대로 확인하겠습니다!!

피드백 주신 대부분의 사항들은 반영 후 커밋하였습니다.
DI 같은 경우는 상단에 코멘트 드린대로 이번주차 작업물 머지 후 통합 브랜치에서 작업할 예정입니다.

아직 많이 부족한 실력이지만 많이 가르쳐주신다면 열심히 배워 성장해보겠습니다! 감사합니다

akuby21 commented:
이번주 프로젝트 진행을 위해 임의로 머지하겠습니다!
혹시 멘토님이 다음부턴 피드백 수정 내역 더블체크 후에 머지하기를 원하신다면 말씀해주세요!
